src/Microsoft.DotNet.Darc/Darc/Operations/VirtualMonoRepo/VmrDiffOperation.cs

```diff
@@ -385,4 +394,117 @@ await repo.ExecuteGitCommand(
 
         return repoPath / VmrInfo.SourceDirName / mapping;
     }
+
+    private async Task CompareSourceTreeWithVmr(Repo sourceRepo, Repo vmrRepo)
+    {
+        var sourceGitClient = _gitRepoFactory.CreateClient(sourceRepo.Remote);
+        var vmrGitClient = _gitRepoFactory.CreateClient(vmrRepo.Remote);
+        var sourceVersionDetails = _versionDetailsParser.ParseVersionDetailsXml(await sourceGitClient.GetFileContentsAsync(VersionFiles.VersionDetailsXml, sourceRepo.Remote, sourceRepo.Ref));
+        var sourceMapping = sourceVersionDetails?.Source?.Mapping ??
+            throw new DarcException($"Product repo {sourceRepo.Remote} is missing source tag in {VersionFiles.VersionDetailsXml}");
+
+        Queue<string?> directoriesToProcess = [];
+        directoriesToProcess.Enqueue(null);
+
+        Dictionary<string, string> fileDifferences = [];
+
+        string? currentPath;
+        while (directoriesToProcess.Count > 0)
+        {
+            currentPath = directoriesToProcess.Dequeue();
+            string vmrMappingPath = $"{VmrInfo.SourcesDir}/{sourceMapping}";
+
+            var sourceFiles = await sourceGitClient.LsTree(sourceRepo.Remote, sourceRepo.Ref, currentPath);
+            var vmrFiles = (await vmrGitClient.LsTree(vmrRepo.Remote, vmrRepo.Ref, $"{vmrMappingPath}{currentPath}"))
+                .Select(entry => (entry.type, entry.sha, path: entry.path.Substring(vmrMappingPath.Length)))
+                .ToList();
+
+            // Handle empty files separately since they all have the same sha
+            (sourceFiles, vmrFiles) = CheckAndFilterEmptyFiles(sourceFiles, vmrFiles, fileDifferences);
+
+            var filesOnlyInVmr = vmrFiles.ToDictionary(file => file.sha, file => file);
+
+            foreach (var sourceFile in sourceFiles)
+            {
+                if (vmrFiles.Any(vmrFile => vmrFile.sha == sourceFile.sha))
+                {
+                    filesOnlyInVmr.Remove(sourceFile.sha);
+                    continue;
+                }
+
+                if (IsBlob(sourceFile.type))
+                {
+                    if (vmrFiles.Any(vmrFile => vmrFile.path == sourceFile.path))
+                    {
+                        fileDifferences[sourceFile.path] = ($"File {sourceFile.path} has different content in the VMR and in the repo");
```

Comment:
I am not a fan of long English sentences that are hard to read through.

Can we do something better here? E.g.
`+ {sourceFile.path}` when the file is added in the target,
`- {sourceFile.path}` when it's missing,
`*` - if content is different?

Possibly would also be nice if the new files is green while the removed one is red but we'd have to allow turning off colouring in case we wanted to use this in automation.

Reply:
Improved, I also kept the shas for files, not sure if that's even valuable tho

Reply:
I removed it, it's just a list of files now with the symbol (-, + or *)